fs/remote/resolver.go

```diff
@@ -132,275 +144,106 @@ func (r *Resolver) resolveFetcher(ctx context.Context, hosts source.RegistryHost
 		return nil, 0, err
 	}
-	if desc.Size == 0 {
-		desc.Size, err = getLayerSize(ctx, hf)
+	if fc.desc.Size == 0 {
+		logger.WithField("ref", fc.refspec.String()).WithField("digest", fc.desc.Digest).
+			Debugf("layer size not found in labels; making a request to remote to get size")
+
+		fc.desc.Size, err = getLayerSize(ctx, hf)
 		if err != nil {
-			return nil, 0, fmt.Errorf("failed to retrieve layer size from %s after it was not found in labels: %w", hf.url, err)
+			return nil, 0, fmt.Errorf("%s from %s: %w", ErrFailedToRetrieveLayerSize, socihttp.RedactHTTPQueryValuesFromString(hf.realBlobURL), err)
 		}
 	}
-	if blobConfig.ForceSingleRangeMode {
+	if r.blobConfig.ForceSingleRangeMode {
 		hf.singleRangeMode()
 	}
-	return hf, desc.Size, err
+	return hf, fc.desc.Size, err
 }
 
-type fetcherConfig struct {
-	hosts      source.RegistryHosts
-	refspec    reference.Spec
-	desc       ocispec.Descriptor
-	maxRetries int
-	minWait    time.Duration
-	maxWait    time.Duration
+type httpFetcher struct {
+	roundTripper  http.RoundTripper
+	scope         string
+	realBlobURL   string
+	baseBlobURL   string
+	urlMu         sync.Mutex
+	digest        digest.Digest
+	singleRange   bool
+	singleRangeMu sync.Mutex
 }
 
 func newHTTPFetcher(ctx context.Context, fc *fetcherConfig) (*httpFetcher, error) {
-	reghosts, err := fc.hosts(fc.refspec)
-	if err != nil {
-		return nil, err
-	}
 	desc := fc.desc
 	if desc.Digest.String() == "" {
-		return nil, fmt.Errorf("digest is mandatory in layer descriptor")
+		return nil, fmt.Errorf("missing digest; a digest is mandatory in layer descriptor")
 	}
 	digest := desc.Digest
-	pullScope, err := repositoryScope(fc.refspec, false)
+
+	pullScope, err := docker.RepositoryScope(fc.refspec, false)
 	if err != nil {
 		return nil, err
 	}
 
 	// Try to create fetcher until succeeded
-	rErr := fmt.Errorf("failed to resolve")
-	for _, host := range reghosts {
+	createFetcherErr := errors.New("")
+	for _, host := range fc.hosts {
 		if host.Host == "" || strings.Contains(host.Host, "/") {
-			rErr = fmt.Errorf("invalid destination (host %q, ref:%q, digest:%q): %w",
-				host.Host, fc.refspec, digest, rErr)
-			continue // Try another
-
+			createFetcherErr = fmt.Errorf("%w (host %q, ref:%q, digest:%q): %w",
+				ErrInvalidHost, host.Host, fc.refspec, digest, createFetcherErr)
+			// Try another
+			continue
 		}
 
-		// Prepare transport with authorization functionality
 		tr := host.Client.Transport
-
-		timeout := host.Client.Timeout
-		if rt, ok := tr.(*rhttp.RoundTripper); ok {
-			rt.Client.RetryMax = fc.maxRetries
-			rt.Client.RetryWaitMin = fc.minWait
-			rt.Client.RetryWaitMax = fc.maxWait
-			rt.Client.Backoff = socihttp.BackoffStrategy
-			rt.Client.CheckRetry = socihttp.RetryStrategy
-			rt.Client.ErrorHandler = socihttp.HandleHTTPError
-			timeout = rt.Client.HTTPClient.Timeout
+		if authClient, ok := tr.(*socihttp.AuthClient); ok {
+			// Get the inner retryable client.
+			retryClient := authClient.Client()
```

Comment:
Ah, right. It would affect **all** subsequent registry operations. This would be OK for span fetches, but for all other registry operations it might be problematic (fetching layers, fetching SOCI artifacts). I think we have a couple options then:

1. Deprecate HTTP knobs in `BlobConfig`
2. Have some sort of global `Blob` client that shares the same authorization with the general global client. I don't see where this would really fit in `RegistryHost`, though. It will have to exist at the process/FS level and I don't think we should inflate FS any more. 
3. Create a new `AuthClient` and `RetryClient` for every layer/blob. Every client will share the same authorization logic, concrete transport (singular connection pool), general retry hooks/functions. We will have a new `AuthClient` and `RetryClient` type in memory for each blob, but I don't think it will matter too much since the client types themselves our fairly light-weight.

Reply:
I went with option 3, as I think that it finds a happy medium between removing the "optimization" entirely and creating additional types that don't really fit the `RegistryHost` pattern we are sticking to. LMK what you think.